sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -441,6 +457,27 @@ object BooleanSimplification extends Rule[LogicalPlan] with PredicateHelper {
 
       case Not(IsNull(e)) => IsNotNull(e)
       case Not(IsNotNull(e)) => IsNull(e)
+
+      // Move `Not` from one side of `EqualTo`/`EqualNullSafe` to the other side if it's beneficial.
+      // E.g. `EqualTo(Not(a), b)` where `b = Not(c)`, it will become
+      // `EqualTo(a, Not(b))` => `EqualTo(a, Not(Not(c)))` => `EqualTo(a, c)`
+      // In addition, `if canSimplifyNot(b)` checks if the optimization can converge
+      // that avoids the situation two conditions are returning to each other.
+      case EqualTo(Not(a), b) if !canSimplifyNot(a) && canSimplifyNot(b) => EqualTo(a, Not(b))
```

Comment:
I wonder how much extra costs will this incur. Since now this transforms the original expression to `EqualTo(a, Not(b))`, we'll need to apply `BooleanSimplification` again to simplify the `Not(b)` (since it is a child of `EqualTo` and the rule runs bottom-up. Does this mean we'd need to run more iterations on the `operatorOptimizationBatch` (see `Optimizer`) until it can reach a fix point?

Reply:
+1, can we write a recursive method to optimize the entire expression tree with one pas?

Reply:
I will think about how I can achieve this

Reply:
I addressed this by using `q.transformExpressionsDownWithPruning()` that optimizes from top to bottom. That is helpful for the case like in the test `'e =!= ('f === ('g === Not('h)))` It will be optimized as
```
'e =!= ('f === ('g === Not('h))) => 'e === ('f =!= ('g === Not('h))) => 'e === ('f === ('g =!= Not('h))) => 'e === ('f === ('g === Not(Not('h)))) => 'e === ('f === ('g === 'h)) 
```
Only the last `Not(Not('h))` will need to re-visit `SimplifyBinaryComparison` one more time

Reply:
Updated